src/libspu/mpc/swift/arithmetic.cc

```diff
@@ -0,0 +1,1871 @@
+// Copyright 2024 Ant Group Co., Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "libspu/mpc/swift/arithmetic.h"
+
+#include <functional>
+#include <iostream>
+#include <string>
+
+#include "libspu/core/type_util.h"
+#include "libspu/core/vectorize.h"
+#include "libspu/mpc/common/communicator.h"
+#include "libspu/mpc/common/prg_state.h"
+#include "libspu/mpc/common/pv2k.h"
+#include "libspu/mpc/swift/hash_func.h"
+#include "libspu/mpc/swift/type.h"
+#include "libspu/mpc/swift/value.h"
+#include "libspu/mpc/utils/ring_ops.h"
+
+namespace spu::mpc::swift {
+
+// note:
+// We don't consider the situation that some party keeps silent
+// We just vote for the TTP, but don't let TTP do the subsequence computation
+// We don't split the offline and online phases
+// We only implement the 3PC protocol of
+// https://eprint.iacr.org/2020/592
+
+NdArrayRef getOrCreateCompactArray(const NdArrayRef& in) {
+  if (!in.isCompact()) {
+    return in.clone();
+  }
+  return in;
+}
+
+// Reference:
+// SWIFT: Super-fast and Robust Privacy-Preserving Machine Learning
+// P6 3.1 Protocol_jmp
+// https://eprint.iacr.org/2020/592.pdf
+NdArrayRef JointMessagePassing(KernelEvalContext* ctx, const NdArrayRef& msg,
+                               size_t rank_send, size_t rank_hash,
+                               size_t rank_recv, std::string_view tag) {
+  auto const field = msg.eltype().as<RingTy>()->field();
+  auto* comm = ctx->getState<Communicator>();
+  auto ty = makeType<RingTy>(field);
+
+  NdArrayRef res(ty, msg.shape());
+  bool inconsistent_bit = false;
+
+  auto rank = comm->getRank();
+  size_t hash_len = 32;
```

Comment:
这种 hardcode/magic number 需要区别下普通变量
```suggestion
  constexpr size_t k_hash_len = 32;
```